Purchases/Misc/PriceFormatterProvider.swift

```diff
@@ -0,0 +1,49 @@
+//
+//  Copyright RevenueCat Inc. All Rights Reserved.
+//
+//  Licensed under the MIT License (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      https://opensource.org/licenses/MIT
+//
+//  PriceFormatterProvider.swift
+//
+//  Created by Juanpe Catalán on 10/3/22.
+
+import Foundation
+
+/// A `NumberFormatter` provider class for prices.
+/// This provider caches the formatter to improve the performance.
+class PriceFormatterProvider {
+
+    private lazy var cachedPriceFormatterForSK1: NumberFormatter = {
+        let formatter = NumberFormatter()
+        formatter.numberStyle = .currency
+        return formatter
+    }()
+
+    func priceFormatterForSK1(withLocale locale: Locale) -> NumberFormatter {
+        if cachedPriceFormatterForSK1.locale != locale {
+            cachedPriceFormatterForSK1.locale = locale
+        }
+        return cachedPriceFormatterForSK1
+    }
+
+    @available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *)
+    private lazy var cachedPriceFormatterForSK2: NumberFormatter = {
+        let formatter = NumberFormatter()
+        formatter.numberStyle = .currency
+        formatter.locale = .autoupdatingCurrent
+        return formatter
+    }()
+
+    @available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *)
+    func priceFormatterForSK2(withCurrencyCode currencyCode: String) -> NumberFormatter {
+        if cachedPriceFormatterForSK2.currencyCode != currencyCode {
+            cachedPriceFormatterForSK2.currencyCode = currencyCode
```

Comment:
This would be changing the currency under the rug of every previously returned formatter. I haven't been following the discussion but I feel like this isn't a desired behavior 😬

Reply:
`NSFormatter` conforms to NSCopying, we could store a copy instead of the currency changes?

Reply:
I see your point 🤔

Well, with this approach each product has its own `priceFormatterProvider`. Therefore, if the currency code is changed, the next time `priceFormatter` is called, the currency code will be updated, and all previously returned formatters will be updated as well. 

My question is if there is a possibility that there are two instances of the same product, one obtained before the currency code changes and another after and we want to show the price using both currency codes, do you know what I mean?

> NSFormatter conforms to NSCopying, we could store a copy instead of the currency changes?
About that, would you store a formatter for each currency code?

Reply:
> This would be changing the currency under the rug of every previously returned formatter. I haven't been following the discussion but I feel like this isn't a desired behavior 😬

Yeah, I hadn't thought of that... You could have an autoUpdatingFormatter or just have it in the docs, but things get tricky: 

In practice, if the storefront changes, you do want the prices to use the new currency, otherwise you'd be presenting potentially really problematic values.

Furthermore, you actually want to entirely drop the old products, because you'll run into consistency problems otherwise:

For example, say that your storefront is USA and you have a product priced at $9.99. 
But you're also selling to Spain, where you're selling it a bit cheaper, at €6.99. 

a) If the storefront updates but the formatter does not, and the old product is still around, the old product stays at price $9.99 when formatted, which is good. 
b) If the storefront updates and the formatter does too, you get €9.99 for the old product, which is incorrect. Then again, the old product is no longer valid in the first place, since when you try to sell it to the customer it'll be priced €6.99, not $9.99. 

So... to truly provide a foolproof experience, we'd have to: 
1. bridge over the storefront changed notification so customers could subscribe to it
2. listen to it and clear all product caches when it fires
3. for bonus points, we could try eagerly re-fetch all the same products we'd cached before, since they're likely (although not guaranteed) to be used in the new storefront too. 

If we don't do 1 and 2 we'll always have a partially broken experience when the storefront changes, we have to kind of pick which broken experience. It does seem like a) is a slightly less broken experience in that at least you'd have outdated products that are internally consistent, so I think we should go with that for now, but we still have to know that it's not an ideal case. 

Storefronts aren't updated often, and chances of it happening in the lifecycle of an app are very low, so I wouldn't call it a huge priority to get this perfect.

Reply:
> In practice, if the storefront changes, you do want the prices to use the new currency, otherwise you'd be presenting potentially really problematic values.

Yes, this is exactly what I thought. 

My question is whether this is the responsibility of the library or the users 🤔 If it's the library, it's something that should be clearly stated in the documentation as expected behaviour, otherwise it could be confusing.

Reply:
> My question is whether this is the responsibility of the library or the users 🤔 If it's the library, it's something that should be clearly stated in the documentation as expected behaviour, otherwise it could be confusing.

I think the library should eventually: 
- delete caches when the storefront is updated
- notify the developer

In the meantime, updating docs to explicitly state that when the storefront is updated our code doesn't respond correctly is a good idea. 

But to be clear about what I meant with my previous comment, 

>  It does seem like a) is a slightly less broken experience in that at least you'd have outdated products that are internally consistent, so I think we should go with that for now, but we still have to know that it's not an ideal case.

I think we should follow @NachoSoto's suggestion for now and return a copy instead of modifying the existing formatter for now. 

Caching is hard. Thanks @NachoSoto for calling this out and @Juanpe for your patience and help in figuring this out! 

I'm going to create tickets for deleting caches when the storefront is updated and adding a method to the purchases delegate.

Reply:
Opened up this one for it https://github.com/RevenueCat/purchases-ios/issues/1382

Reply:
But the way this is solved in this PR requires calling the method one more time, which is more of an accident than the framework explicitly listening to and reacting to changes.

Reply:
So...the idea is to return a copy when the currency code is different. My question is, in this case, we should replace the cached formatter with the new copy, right?

Reply:
Yeah exactly, store and return the new copy.

Reply:
Ok, as we discussed I've modified the logic. Now, we'll store and return a copy, if the currency code is different :)

Here is the commit: https://github.com/RevenueCat/purchases-ios/pull/1374/commits/d9dda34c7a211fc17a516693700c700a146c95ef